src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -105,14 +125,36 @@ public static class ActivatorUtilities
                 }
             }
 
-            Type?[] argumentTypes = new Type[parameters.Length];
-            for (int i = 0; i < argumentTypes.Length; i++)
+            Type?[] argumentTypes;
+            if (parameters.Length == 0)
             {
-                argumentTypes[i] = parameters[i]?.GetType();
+                argumentTypes = Type.EmptyTypes;
+            }
+            else
+            {
+                argumentTypes = new Type[parameters.Length];
+                for (int i = 0; i < argumentTypes.Length; i++)
+                {
+                    argumentTypes[i] = parameters[i]?.GetType();
+                }
             }
 
             FindApplicableConstructor(instanceType, argumentTypes, out ConstructorInfo constructorInfo, out int?[] parameterMap);
-            var constructorMatcher = new ConstructorMatcher(constructorInfo);
+
+            // Find the ConstructorInfoEx from the given constructorInfo.
+            constructor = null;
+            foreach (ConstructorInfoEx ctor in constructors)
```

Comment:
Do we need to worry about a case where there are a lot of constructors and this O(n) lookup becomes problematic?